torch/distributed/_cuda_p2p/__init__.py

```diff
@@ -121,3 +121,360 @@ def get_p2p_buffer_size(group: c10d.ProcessGroup) -> int:
     extended_api=True,
     devices=["cuda"],
 )
+
+
+_test_with_non_cuda_p2p_group: bool = False
+
+
+@contextmanager
+def test_with_non_cuda_p2p_group():
+    """
+    Force ops in this file to work with non-cuda_p2p groups for testing
+    purposes. Not thread safe.
+    """
+    global _test_with_non_cuda_p2p_group
+    prev = _test_with_non_cuda_p2p_group
+    try:
+        _test_with_non_cuda_p2p_group = True
+        yield
+    finally:
+        _test_with_non_cuda_p2p_group = prev
+
+
+def _pipelined_all_gather_and_consume(
+    shard: torch.Tensor,
+    shard_consumer: Callable[[torch.Tensor, int], None],
+    ag_out: torch.Tensor,
+    group: c10d.ProcessGroup,
+) -> None:
+    """
+    Perform the following logic with micro-pipelined computation and
+    communication:
+
+        tensor = all_gather_tensor(shard, gather_dim=1, group=group)
+        chunks = tensor.chunk(group.size())
+        for src_rank, chunk in enumerate(chunks):
+            shard_consumer(chunk, src_rank)
+
+    NOTE:
+    - The shard passed to shard consumer will always be contiguous.
+    """
+    p2p_buf_sz_req = shard.numel() * shard.element_size()
+    if get_p2p_buffer_size(group) < p2p_buf_sz_req:
+        # We preferred the caller to handle fallback so that the computation
+        # doesn't need to be decomposed.
+        raise RuntimeError(
+            f"_pipelined_all_gather_and_consume on input with shape={shard.shape} "
+            f"and dtype={shard.dtype} requires {p2p_buf_sz_req} bytes of p2p buffers "
+            f"(got {get_p2p_buffer_size(group)} bytes)."
+        )
+
+    backend = get_cuda_p2p_backend(group)
+    group_size = group.size()
+    rank = group.rank()
+
+    backend.stream().wait_stream(torch.cuda.current_stream())
+    local_p2p_buf = backend.get_p2p_buffer(rank, shard.shape, shard.dtype)
+
+    chunks = ag_out.chunk(group.size())
+
+    # While consuming local shard, copy it to the local p2p buffer
+    # in another stream.
+    shard_consumer(shard, rank)
+    chunks[rank].copy_(shard)
+
+    with torch.cuda.stream(backend.stream()):
+        local_p2p_buf.copy_(shard)
+        work = backend.intra_node_barrier()
```

Comment:
Why is this the only one that needs a barrier?

Reply:
After this barrier, the condition "shard `i` is available in rank `i`'s p2p buffer" will be met, after which each rank can "pull" from remote ranks without additional synchronization. We have another barrier at the end, which ensures that "all ranks have finished reading from remote ranks" so that the p2p buffers can be safely reused.